dissect/hypervisor/backup/vbk.py

```diff
@@ -0,0 +1,1044 @@
+# References:
+# - Veeam extract utility
+# - Veeam agent
+from __future__ import annotations
+
+from functools import cached_property, lru_cache
+from io import BytesIO
+from typing import BinaryIO, Generic, Iterator, Optional, TypeVar
+from zlib import crc32
+
+from dissect.cstruct import Structure
+from dissect.util.crc32c import crc32c
+from dissect.util.stream import AlignedStream
+from dissect.util.xmemoryview import xmemoryview
+
+try:
+    from lz4.block import decompress as lz4_decompress
+except ImportError:
+    from dissect.util.compression.lz4 import decompress as lz4_decompress
+
+from dissect.hypervisor.backup.c_vbk import c_vbk
+from dissect.hypervisor.exceptions import Error
+
+PAGE_SIZE = 4096
+"""VBK page size."""
+
+
+class VBKError(Error):
+    pass
+
+
+class NotAFileError(VBKError):
+    pass
+
+
+class NotADirectoryError(VBKError):
+    pass
+
+
+class VBK:
+    """Veeam Backup (VBK) file implementation.
+
+    Args:
+        fh: The file handle of the VBK file to read.
+        verify: Whether to verify checksums.
+
+    References:
+        - CMeta
+        - CStgFormat
+
+    Notes:
+        - **TODO**: Encryption
+        - **TODO**: Incrememental backups
+    """
+
+    def __init__(self, fh: BinaryIO, verify: bool = True):
+        self.fh = fh
+
+        fh.seek(0)
+        self.header = c_vbk.StorageHeader(fh)
+
+        self.format_version = self.header.FormatVersion
+        self.block_size = self.header.StandardBlockSize
+
+        # First slot starts at PAGE_SIZE because StorageHeader is considered to be PAGE_SIZE large
+        self.slot1 = SnapshotSlot(self, PAGE_SIZE)
+        # Second slot starts at PAGE_SIZE + slot1 size
+        self.slot2 = SnapshotSlot(self, PAGE_SIZE + self.slot1.size)
+
+        populated_slots = filter(lambda slot: slot.header.ContainsSnapshot, (self.slot1, self.slot2))
+
+        if verify:
+            populated_slots = filter(lambda slot: slot.verify(), populated_slots)
+
+        if not (active_slot := max(populated_slots, key=lambda slot: slot.descriptor.Version, default=None)):
+            raise VBKError("No active VBK metadata slot found")
+
+        self.active_slot: SnapshotSlot = active_slot
+
+        self.root = RootDirectory(
+            self,
+            self.active_slot.descriptor.DirectoryRoot.RootPage,
+            self.active_slot.descriptor.DirectoryRoot.Count,
+        )
+        self.block_store = MetaVector(
+            self,
+            StgBlockDescriptorV7 if self.is_v7 else StgBlockDescriptor,
+            self.active_slot.descriptor.BlocksStore.RootPage,
+            self.active_slot.descriptor.BlocksStore.Count,
+        )
+
+    def is_v7(self) -> bool:
+        return self.format_version == 7 or self.format_version == 0x10008 or self.format_version >= 9
```

Comment:
Yes, it's named according to how Veeam themselves name this check. And unsure, the logic should be similar, otherwise Veeam tools themselves should also have issues :smile: only time will tell.